core/src/main/java/com/datastrato/gravitino/storage/relational/mapper/FilesetVersionMapper.java

```diff
@@ -97,4 +100,33 @@ void insertFilesetVersionOnDuplicateKeyUpdate(
           + " SET deleted_at = UNIX_TIMESTAMP(CURRENT_TIMESTAMP(3)) * 1000.0"
           + " WHERE fileset_id = #{filesetId} AND deleted_at = 0")
   Integer softDeleteFilesetVersionsByFilesetId(@Param("filesetId") Long filesetId);
+
+  @Delete(
+      "DELETE FROM "
+          + VERSION_TABLE_NAME
+          + " WHERE deleted_at > 0 AND deleted_at < #{legacyTimeLine} LIMIT #{limit}")
+  Integer deleteFilesetVersionsByLegacyTimeLine(
+      @Param("legacyTimeLine") Long legacyTimeLine, @Param("limit") int limit);
+
+  @Select(
+      "SELECT fileset_id as filesetId,"
+          + " Max(version) as version, deleted_at as deletedAt,"
+          + " metalake_id as metalakeId, catalog_id as catalogId, schema_id as schemaId,"
+          + " storage_location as storageLocation"
+          + " FROM "
+          + VERSION_TABLE_NAME
+          + " WHERE version > #{versionRetentionCount} AND deleted_at = 0"
+          + " GROUP BY fileset_id")
+  List<FilesetVersionPO> selectFilesetVersionsByRetentionCount(
```

Comment:
Maybe we can use a `List<Long, Long>` here just to get the `fileset_id` and `version` fields to reduce the data size returned by MySQL consider that this will scan the entire table.

Reply:
good catch :-)  👍